data/transactions/transaction.go

```diff
@@ -343,12 +343,12 @@ func (tx Transaction) WellFormed(spec SpecialAddresses, proto config.ConsensusPa
 			return fmt.Errorf("tx.ForeignAssets too long, max number of foreign assets is %d", proto.MaxAppTxnForeignAssets)
 		}
 
-		if len(tx.ApprovalProgram) > proto.MaxAppProgramLen {
-			return fmt.Errorf("approval program too long. max len %d bytes", proto.MaxAppProgramLen)
+		if len(tx.ApprovalProgram) > ((1 + tx.ExtraProgramPages) * proto.MaxAppProgramLen) {
```

Comment:
usage of `tx.ExtraProgramPages` needs to be under a feature switch. Consider a scenario:
1. Half of nodes upgraded to `ExtraProgramPages`, other half is not.
2. Nodes receive a txn with `ExtraProgramPages` = 1 and `len(ApprovalProgram) = proto.MaxAppProgramLen * 1.5`. Then upgrade nodes will pass the txn even before the protocol upgrade, and other will break (most likely on txn decoding)

Reply:
Until the protocol upgrade, having a ExtraProgramPages > 0 will be rejected either way.  An upgraded node will be mad that ExtraProgramPages > 0 (since proto.MaxExtraAppProgramPages will be initialized to 0), while an unupgraded node will be made it contains the field at all. So I think the protocol upgrade itself acts as the feature flag.

ExtraProgramPages = 0 would be trouble, except we will reject that because it should be omitted when 0.

Reply:
correct if moving `if tx.ExtraProgramPages > proto.MaxExtraAppProgramPages` check right above.
I assume old code will fail on decoding `ExtraProgramPages` but it is better to double check.

Reply:
> ExtraProgramPages = 0 would be trouble, except we will reject that because it should be omitted when 0.

@jannotti could you elaborate? 0 looks ok here.

Reply:
We usually use 'omitempty' (and certainly should here, for the extra pages parameter) and when we do that, I think it becomes impossible to sign a transaction with 0 in it for that field, because we must do some check where we reencode to canonical representation (which will differ, since the 0 field will be dropped) and check the signature against that.

Reply:
in that case, a program of len <= proto.MaxAppProgramLen will get dropped?

Reply:
No, I was just pointing out that @algorandskiy 's concern might occur if, before the protocol upgrade:
a) Node with new release accepts a txn containing extrapages=0
b) Not upgraded node can't understand it.

But, that's not actually possible.  Txns can't *explicitly* set the extra pages field to 0. Instead, they omit it if they mean 0. So I don't think anything bad can happen.

Reply:
still I think we are fine:
1. old node receives a txn with omitted `ExtraProgramPages` => it decodes and proceeds as now
2. new node creates a txn without `ExtraProgramPages` => fine
3. new node receives a txn with omitted`ExtraProgramPages` => decodes as ExtraProgramPages=0 and proceeds with new code checks
4. new node receives a txn with non-zero `ExtraProgramPages` => fine, execute new code checks

Reply:
I agree

Reply:
I agree. So move the check `tx.ExtraProgramPages > proto.MaxExtraAppProgramPages` above and write a test checking
1) pre-upgrade txn with ExtraProgramPages=1 fails
2) post-upgrade txn with ExtraProgramPages=1 succeed